dotCMS/src/main/java/com/dotcms/security/secret/SecretsStoreKeyStoreImpl.java

```diff
@@ -0,0 +1,352 @@
+        this(getSecretStorePath());
+    }
+
+    /**
+     * returns the password to the store - it can be set as a config variable (which can read from an
+     * environment or system property) or will default to the SHA-256 digest of the clusterId
+     * 
+     * @return
+     */
+    @VisibleForTesting
+    private char[] loadStorePassword() {
+        return getFromCache(SECRETS_KEYSTORE_PASSWORD_KEY,
+                () -> Config.getStringProperty(SECRETS_KEYSTORE_PASSWORD_KEY, digest(ClusterFactory.getClusterId()))
+                                .toCharArray());
+
+    }
+
+    /**
+     * This will create a Keystore file for reading/writing if there is not one already there,
+     * otherwise, it will return the one there
+     * 
+     * @return
+     */
+    private File createStoreIfNeeded() {
+        final File secretStoreFile = new File(secretsKeyStorePath);
+        if (!secretStoreFile.exists()) {
+            try {
+                final KeyStore keyStore = KeyStore.getInstance(SECRETS_STORE_KEYSTORE_TYPE);
+                keyStore.load(null, loadStorePassword());
+                flushCache();
+                saveSecretsStore(keyStore);
+            } catch (Exception e) {
+                Logger.warn(this.getClass(), "unable to create secrets store " + SECRETS_STORE_FILE + ": " + e);
+                throw new DotRuntimeException(e);
+            }
+        }
+        return secretStoreFile;
+    }
+
+
+    /**
+     * loads up the Keystore from disk
+     * 
+     * @return
+     */
+    private KeyStore getSecretsStore() {
+        final File secretStoreFile = createStoreIfNeeded();
+        try {
+            final KeyStore keyStore = KeyStore.getInstance(SECRETS_STORE_KEYSTORE_TYPE);
+            try (InputStream inputStream = Files.newInputStream(secretStoreFile.toPath())) {
+                keyStore.load(inputStream, loadStorePassword());
+            }
+
+            return keyStore;
+
+        } catch (Exception e) {
+            Logger.warn(this.getClass(), "unable to load secrets store " + SECRETS_STORE_FILE + ": " + e);
+            throw new DotRuntimeException(e);
+        }
+
+    }
+
+    /**
+     * Persists the keystore, tries to do a 2 phase commit, saving to a tmp file, copying that to the
+     * keystore file, then deleting the tmp file
+     * 
+     * @param keyStore
+     * @return
+     */
+    private KeyStore saveSecretsStore(final KeyStore keyStore) {
+        final File secretStoreFile = new File(secretsKeyStorePath);
+        final File secretStoreFileTmp = new File(secretStoreFile.getParent(), "dotSecretsStore_" + System.currentTimeMillis() + ".p12.tmp");
+
+        secretStoreFileTmp.getParentFile().mkdirs();
+        try (OutputStream fos = Files.newOutputStream(secretStoreFileTmp.toPath())) {
+            keyStore.store(fos, loadStorePassword());
+            FileUtil.copyFile(secretStoreFileTmp, secretStoreFile);
+            secretStoreFileTmp.delete();
```

Comment:
you can create a temporal file by using FileUtil.createTemporalFile instead of handling by yourself.
If you need to do this way anyhow, you could probably add a delete in a finally sentence